src/main/java/com/commercetools/config/ctpTypes/AggregatedCtpTypesValidationResult.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    /**
+     * @return aggregated result where neither error found, either types were updated
+     */
+    @Nonnull
+    public static AggregatedCtpTypesValidationResult ofEmpty() {
+        return new AggregatedCtpTypesValidationResult(null, null);
+    }
+
+    /**
+     * @param aggregatedErrorMessage complete message to display to customer in the logs
+     * @return instance with error message
+     */
+    @Nonnull
+    public static AggregatedCtpTypesValidationResult ofErrorMessage(@Nonnull String aggregatedErrorMessage) {
+        requireNonBlank(aggregatedErrorMessage, "aggregatedErrorMessage argument must be non-blank");
+        return new AggregatedCtpTypesValidationResult(aggregatedErrorMessage, null);
+    }
+
+    /**
+     * @param updatedTypes list of tuples, where pair key(left) is a tenant name, value(right) - list of updated types
+     *                     on this tenant.
+     * @return instance with update tenant-types pairs
+     */
+    @Nonnull
+    public static AggregatedCtpTypesValidationResult ofUpdatedTypes(@Nonnull List<Pair<String, List<Type>>> updatedTypes) {
+        requireNonEmpty(updatedTypes);
+        return new AggregatedCtpTypesValidationResult(null, updatedTypes);
+    }
+
+    /**
+     * From the list of all {@link TenantCtpTypesValidationAction}:<ul>
+     * <li>if the {@code tenantTypeActions} list has at least one error: return error aggregated result</li>
+     * <li>otherwise, if the {@code tenantTypeActions} list at least one update action: execute the actions and return
+     * aggregated result with the list of {@link #updatedTypes}</li>
+     * <li>otherwise return empty object</li>
+     * </ul>
+     *
+     * @param tenantTypeActions list of tenant type actions to aggregate
+     * @return completion stage with {@link AggregatedCtpTypesValidationResult} containing the result of the
+     * errors/update actions executing and aggregating.
+     */
+    @Nonnull
+    public static CompletionStage<AggregatedCtpTypesValidationResult> executeAndAggregateTenantValidationResults(
+            @Nonnull List<TenantCtpTypesValidationAction> tenantTypeActions) {
+
+        List<TenantCtpTypesValidationAction> typeErrors = tenantTypeActions.stream()
+                .filter(TenantCtpTypesValidationAction::hasErrors)
+                .collect(toList());
+
+        // 1. If some errors exist - return error result
+        if (typeErrors.size() > 0) {
+            return completedFuture(
+                    ofErrorMessage(format("CTP Types configuration exception:%n%s%n",
+                            typeErrors.stream()
+                                    .flatMap(validationError -> validationError.getErrorMessages().stream())
+                                    .map(errorMessage -> "    " + errorMessage)
+                                    .collect(joining(format("%n"))))));
+
+        }
+
+        List<TenantCtpTypesValidationAction> typeUpdates = tenantTypeActions.stream()
```

Comment:
Wouldn't it be more efficient instead of streaming twice here and [here](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/107/files#diff-b8afde1f1f8da6be73b8fe58f3f019aaR122) for getting list of errors and the list of updates, to do it in one traversal classical loop, and add to either errors or updates accordingly?

Reply:
yes, from _efficiency_ perspective it is better, and it was like this from the beginning, but then i sacrificed _efficiency_ for the sake of  _readability_, so the code looks split and easier to read. 

In current case, having only tens of items efficiency impact is not so significant.

Reply:
ok, I guess in this case the number of update actions and/or errors isn't/wont be that significant anyways.